flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesStateHandleStoreTest.java

```diff
@@ -725,4 +856,176 @@ public void testRemoveAllHandles() throws Exception {
+                                    is(notNullValue(RetrievableStateHandle.class)));
+
+                            // First remove attempt should fail when we're discarding the underlying
+                            // state.
+                            final Exception exception =
+                                    assertThrows(
+                                            Exception.class, () -> store.releaseAndTryRemove(key));
+                            assertThat(exception, FlinkMatchers.containsCause(discardException));
+
+                            // Now we should see that the node is "soft-deleted". This means it can
+                            // no longer be accessed by the get methods, but the underlying state
+                            // still exists.
+                            assertThat(store.getAllAndLock().size(), is(0));
+                            assertThrows(Exception.class, () -> store.getAndLock(key));
+                            assertThat(TestingLongStateHandleHelper.getGlobalDiscardCount(), is(0));
+                            assertThat(getLeaderConfigMap().getData().containsKey(key), is(true));
+
+                            // Second retry should succeed and remove the underlying state and the
+                            // reference in config map.
+                            assertThat(store.releaseAndTryRemove(key), is(true));
+                            assertThat(store.getAllAndLock().size(), is(0));
+                            assertThrows(Exception.class, () -> store.getAndLock(key));
+                            assertThat(TestingLongStateHandleHelper.getGlobalDiscardCount(), is(1));
+                            assertThat(getLeaderConfigMap().getData().containsKey(key), is(false));
+                        });
+            }
+        };
+    }
+
+    @Test
+    public void testReleaseAndTryRemoveAllIsIdempotent() throws Exception {
+        new Context() {
+            {
+                runTest(
+                        () -> {
+                            leaderCallbackGrantLeadership();
+
+                            final KubernetesStateHandleStore<
+                                            TestingLongStateHandleHelper.LongStateHandle>
+                                    store =
+                                            new KubernetesStateHandleStore<>(
+                                                    flinkKubeClient,
+                                                    LEADER_CONFIGMAP_NAME,
+                                                    longStateStorage,
+                                                    filter,
+                                                    LOCK_IDENTITY);
+                            final int numKeys = 10;
+
+                            final RuntimeException discardException =
+                                    new RuntimeException("Test exception.");
+                            for (int idx = 0; idx < numKeys; idx++) {
```

Comment:
By adding a handle that doesn't fail during discard we could verify the behavior for these as well. This goes into the direction which I mentioned earlier whether we remove each entry individually based on its discard call being successful or not (how it's done in the ZK implementation) or whether we let none be deleted as soon as a single entry has an error during discard (how it's currently implemented in k8s)

Reply:
I've made every odd state to fail.